### ISSUE TITLE
fancy repetition

### ISSUE THREAD

**Issue by nrc** [dtolnay/quote#2]
I think this is a feature request, but it might be that I just couldn't figure out how to do it.

Inside repetition expressions, I'd like to be able to use more than just an identifier, e.g., 

```
let foo: Vec<_> = ...;
let bar: Vec<_> = ...;
quote! {
    #(foo: bar),*
}
```

Should produce:

```
foo0: bar0, foo1: bar1, ....
```

**Comment by dtolnay** [comment]
Yeah ideally it would work with all the same tricks as macros-by-example but for now `quote!` itself is a macro-by-example so it is pretty limited. Here is one workaround:

```rust
let foo = vec!["foo0", "foo1"];
let bar = vec!["bar0", "bar1"];
let pairs = foo.into_iter().zip(bar).map(|(f, b)| quote!(#f: #b));
quote! {
    #(pairs),*
}
```

I will think about ways to make this better.

**Comment by dtolnay** [comment]
I got this closure-like syntax working:

```rust
quote! {
    #(|foo, bar| #foo: #bar),*
}
```

Let's see if we can do any better.

**Comment by dtolnay** [comment]
@nrc I published [this PR](https://github.com/dtolnay/quote/pull/3) as [0.3.0-rc1](https://crates.io/crates/quote/0.3.0-rc1). Try it out and let me know what you think. It supports fancy repetitions:

```rust
quote! {
    #(#a: #b),*

    #(
        let #var = #value;
    )*

    #(
        #(#into_iter_of_into_iters)*
    )*

    // caveat: breaking change for non-fancy repetitions
    #(#e)*
}
```

It should work basically just like `macro_rules!` except that you cannot mix iterators and non-iterators at the same level of repetition, so if `a` is an IntoIterator and `b` is not then you cannot do `#(#a #b)*`.

**Comment by dtolnay** [comment]
Fixed in [0.3.0](https://github.com/dtolnay/quote/releases/tag/0.3.0).
